ibc-testkit/src/testapp/ibc/core/types.rs

```diff
@@ -39,69 +48,106 @@ use crate::testapp::ibc::clients::mock::client_state::{
+    pub channel_end_store: ProtobufStore<SharedStore<S>, ChannelEndPath, ChannelEnd, RawChannelEnd>,
+    /// A typed-store for send sequences
+    pub send_sequence_store: JsonStore<SharedStore<S>, SeqSendPath, Sequence>,
+    /// A typed-store for receive sequences
+    pub recv_sequence_store: JsonStore<SharedStore<S>, SeqRecvPath, Sequence>,
+    /// A typed-store for ack sequences
+    pub ack_sequence_store: JsonStore<SharedStore<S>, SeqAckPath, Sequence>,
+    /// A typed-store for packet commitments
+    pub packet_commitment_store: BinStore<SharedStore<S>, CommitmentPath, PacketCommitment>,
+    /// A typed-store for packet receipts
+    pub packet_receipt_store: TypedSet<SharedStore<S>, ReceiptPath>,
+    /// A typed-store for packet ack
+    pub packet_ack_store: BinStore<SharedStore<S>, AckPath, AcknowledgementCommitment>,
+    /// IBC Events
+    pub events: Arc<Mutex<Vec<IbcEvent>>>,
+    /// message logs
+    pub logs: Arc<Mutex<Vec<String>>>,
+}
 
-    /// Emitted IBC events in order
-    pub events: Vec<IbcEvent>,
+impl<S> MockIbcStore<S>
+where
+    S: ProvableStore + Debug,
+{
+    pub fn new(store: S) -> Self {
+        let shared_store = SharedStore::new(store);
+        Self {
+            client_counter: Arc::new(Mutex::new(0)),
+            conn_counter: Arc::new(Mutex::new(0)),
+            channel_counter: Arc::new(Mutex::new(0)),
+            client_processed_times: Arc::new(Mutex::new(Default::default())),
+            client_processed_heights: Arc::new(Mutex::new(Default::default())),
+            consensus_states: Arc::new(Mutex::new(Default::default())),
+            client_state_store: TypedStore::new(shared_store.clone()),
+            consensus_state_store: TypedStore::new(shared_store.clone()),
+            connection_end_store: TypedStore::new(shared_store.clone()),
+            connection_ids_store: TypedStore::new(shared_store.clone()),
+            channel_end_store: TypedStore::new(shared_store.clone()),
+            send_sequence_store: TypedStore::new(shared_store.clone()),
+            recv_sequence_store: TypedStore::new(shared_store.clone()),
+            ack_sequence_store: TypedStore::new(shared_store.clone()),
+            packet_commitment_store: TypedStore::new(shared_store.clone()),
+            packet_receipt_store: TypedStore::new(shared_store.clone()),
+            packet_ack_store: TypedStore::new(shared_store.clone()),
+            events: Arc::new(Mutex::new(Vec::new())),
+            logs: Arc::new(Mutex::new(Vec::new())),
+            store: shared_store,
+        }
+    }
+}
 
-    /// Logs of the IBC module
-    pub logs: Vec<String>,
+impl<S> Default for MockIbcStore<S>
+where
+    S: ProvableStore + Debug + Default,
+{
+    fn default() -> Self {
+        Self::new(S::default())
+    }
 }
 
 /// A context implementing the dependencies necessary for testing any IBC module.
 #[derive(Debug)]
-pub struct MockContext {
+pub struct MockGenericContext<S>
```

Comment:
Sorry that I forgot to mention this. We should make it `MockContext`. The idea was to avoid noise in the tests and use a concrete type `MockContext`, which concretizes the generic type `S`.

Reply:
We can directly use `MockContext<S = MyBasecoinStore>`.

Reply:
No worries. Thanks. Let's adjust naming in a later PR when everything's settled.